role based access control guide (#21)
biscuit-auth/website: content/docs/guides/RBAC.md

+++
title = "Role based access control"
description = "Implementing RBAC with Biscuit"
date = 2021-05-01T08:00:00+00:00
updated = 2021-05-01T08:00:00+00:00
draft = false
weight = 10
sort_by = "weight"
template = "docs/page.html"

[extra]
lead = "Implementing RBAC with Biscuit"
toc = false
top = false
+++

Role-based access control is a common authorization model where a set of permissions is assigned to a role, and a user or program can have one or more roles.

> This makes permissions more manageable than giving them to users directly: a role can be designed for a set of tasks, and can be given or taken back from the user depending on their duties, in one operation, instead of reviewing te user's entire set of rights. Changing the permissions of a role can also be done without going through all the users.

Reviewer: ```suggestion
This makes permissions more manageable than giving them to users directly: a role can be designed for a set of tasks, and can be given or taken back from the user depending on their duties, in one operation, instead of reviewing the user's entire set of rights. Changing the permissions of a role can also be done without going through all the users.
```

## Example

> Let us imagine a space faring package delivery company. Each member of the company has specific duties, represented by roles, that can perform specific actions.

Reviewer: ```suggestion
Let us imagine a space-faring package delivery company. Each member of the company has specific duties, represented by roles, that can perform specific actions.
```

// accountants can check the billing info and the address for invoicing
role("accounting", ["billing:read", "billing:write", "address:read"]);
// support people can update delivery info
role("support", ["address:read", "address:write"]);
// the pilot can drive and learn the delivery address
role("pilot", ["spaceship:drive", "address:read"]);
// delivery people can learn the address and handle the package
role("delivery", ["address:read", "package:load", "package:unload", "package:deliver"]);

// associate users to roles
// this would represent a database table holding both user data and roles,
// but similar facts could be derived from a join table between User and Role tables
user_roles(0, "Professor Farnsworth", ["admin"]);
user_roles(1, "Hermes Conrad", ["accounting"]);
user_roles(2, "Amy Wong", ["support"]);
user_roles(3, "Leela", ["pilot", "delivery"]);
user_roles(4, "Fry", ["delivery"]);
{% end %}

We want to check if an operation is authorized, depending on the user requesting it. Typically, the user id would be carried in a fact like`user(0)`, in the first block of a Biscuit token. Each employee gets issued their own token.

> From that user id, we would look up in the database the user's roles, and for each roles the authorized operations, and load that as facts. We can then check that we have the rights to perform the operation:

Reviewer: ```suggestion
From that user id, we would look up in the database the user's roles, and for each role the authorized operations, and load that as facts. We can then check that we have the rights to perform the operation:
```

{% datalog() %}
role("admin", ["billing:read", "billing:write", "address:read", "address:write"] );
role("accounting", ["billing:read", "billing:write", "address:read"]);
role("support", ["address:read", "address:write"]);
role("pilot", ["spaceship:drive", "address:read"]);
role("delivery", ["address:read", "package:load", "package:unload", "package:deliver"]);

user_roles(0, "Professor Farnsworth", ["admin"]);
user_roles(1, "Hermes Conrad", ["accounting"]);
user_roles(2, "Amy Wong", ["support"]);
user_roles(3, "Leela", ["pilot", "delivery"]);
user_roles(4, "Fry", ["delivery"]);

// we got this from a cookie or Authorization header
user(1);
// we know from the request which kind of operation we want
operation("billing:write");

// we materialize the rights
right($id, $principal, $operation) <-
  user($id),
  operation($operation),
  user_roles($id, $principal, $roles),
  role($role, $permissions),
  $roles.contains($role),
  $permissions.contains($operation);

allow if
  operation($op),
  right($id, $principal, $op);

deny if true;
{% end %}

> Why are we loading data from the database and checking the rights here, while we could do all of that sa part of a SQL query? After all, Datalog is doing similar work, joining facts like we would join tables.

Reviewer: ```suggestion
Why are we loading data from the database and checking the rights here, while we could do all of that as part of a SQL query? After all, Datalog is doing similar work, joining facts like we would join tables.
```

user(3);

// the token is attenuated with a new block containing those checks
check if
  resource("high priority"),
  operation($op),
  role("high priority", "delivery", $permissions),
  $permissions.contains($op);
check if
  time($date),
  $date < 3000-01-31T12:00:00.00Z;

// data from the request
operation("address:read");
resource("high priority");
// provided by the authorizer
time(3000-01-31T11:00:00.00Z);

// user roles loaded from the user id in the first block
user_roles(3, "Leela", "high priority", ["pilot", "delivery"]);

// roles loaded from the ressource and the list from user_roles
role("high priority", "pilot", ["spaceship:drive", "address:read"]);
role("high priority", "delivery", ["address:read", "package:load", "package:unload", "package:deliver"]);

// we materialize the rights
right($id, $principal, $operation, $priority) <-
  user($id),
  operation($operation),
  resource($priority),
  user_roles($id, $principal, $priority, $roles),
  role($priority, $role, $permissions),
  $roles.contains($role),
  $permissions.contains($operation);

allow if
  operation($op),
  resource($priority),
  right($id, $principal, $op, $priority);

deny if true
{% end %}

> Attenuating a token does not increase rights: if suddenly Leela loses the delivery role, the chec of the attenuated token could succeed but authorization would fail both for Leela and Bender because the `right` fact will not be generated.

Reviewer: ```suggestion
Attenuating a token does not increase rights: if suddenly Leela loses the delivery role, the check of the attenuated token could succeed but authorization would fail both for Leela and Bender because the `right` fact would not be generated.
```
